sophia/pa-2807 limit core autodetect to 16

SophiaSR opened the pull request:
Semgrep's memory usage grows linearly with the number of jobs. For now, limiting --jobs to 16 when autodetected. Otherwise, we can get into a situation where we detect a large number of underlying CPU cores, but the customer hasn't allocated a corresponding amount of memory.

Test Plan: For now tested by setting the max cores to 4, on a 10 core machine. Checked to see that this would run with 4 when no value given for jobs, but would run as 10 when set explicitly to 10 by user. 

PR checklist:

- [ ] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [ ] Tests included or PR comment includes a reproducible test plan
- [ ] Documentation is up-to-date
- [ ] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [ ] Change has no security implications (otherwise, ping security team)

If you're unsure about any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

SophiaSR committed:
sophia/pa-2807 limit core autodetect to 16 since currently autodetect can detect a very large number of cores, leading to problems on large machines. This fixes that to limit it to 16

brandonspark commented:
So *you* got the perf work.

aryx commented on cli/src/semgrep/engine.py:
```diff
@@ -63,7 +63,8 @@ def get_cpu_count() -> int:
         if self == EngineType.PRO_INTERFILE:
             return 1
-        return self.get_cpu_count()
+
```
Please add a comment in the code why we use a max of 16 (mostly copy what you put in your PR explanation in the code instead).

aryx reviewed (changes requested):
First PR!

Please can you also update osemgrep code, because this python code will at some point be removed. The relevant code is in src/osemgrep/cli_scan/Scan_CLI.ml, look for the call to get_cpu_count().

SophiaSR committed:
sophia/pa-2807 added comments and updated osemgrep code as well to include the same maximum

SophiaSR commented:
Added comments, and updated the osemgrep code as well. Not sure why the lint/pre-commit test is failing though, could someone take a look at that or point me in the right direction?

aryx commented:
hmm, I get the same error in another PR. Something is broken on develop probably.

aryx commented on cli/src/semgrep/engine.py:
```diff
@@ -63,7 +63,8 @@ def get_cpu_count() -> int:
         if self == EngineType.PRO_INTERFILE:
             return 1
-        return self.get_cpu_count()
+        """Maxing out number of cores used to 16 if more not requested to not overload on large machines"""
```
I think in Python those docstring comments are usually written to explain the whole function and put
just after the ':'.
I think here you should use the '#' syntax for this kind of comment

aryx commented:
This is a user-facing change, so you should add an entry in changelog.d/pa-2807.changed

aryx reviewed (changes requested):
Mostly good

aryx commented:
Hmm let me rerun the failing lint / pre-commit workflow.

aryx commented:
seems to work now.

aryx commented:
(Github action is quite flaky, just FYI, so it's not unusual we need to relaunch manually some failing jobs).

SophiaSR committed:
sophia/pa-2807 changing to correct python comment syntax

SophiaSR committed:
sophia/2807 adding entry in changelog

aryx commented:
This is a user-facing change, so you should add an entry in changelog.d/[pa-2807](https://linear.app/r2c/issue/PA-2807).changed

SophiaSR commented:
Added entry in changelog and updated to correct syntax

emjin reviewed (approved):
🚀

IagoAbal commented on cli/src/semgrep/engine.py:
```diff
@@ -63,7 +63,8 @@ def get_cpu_count() -> int:
         if self == EngineType.PRO_INTERFILE:
             return 1
-        return self.get_cpu_count()
+        # Maxing out number of cores used to 16 if more not requested to not overload on large machines
```
If I didn't know what happened here, and I just found this comment, I would not know what "overload" means here.

IagoAbal commented:
I left a comment in Linear too, but this solution seems too ad-hoc. And I'm sure those using larger machines with the proper amount of memory will not realize that we have this limit, it's a bit like `interfile: true`, you know it's there but it's so easy to forget about it.

IagoAbal commented on cli/src/semgrep/engine.py:
```diff
@@ -63,7 +63,8 @@ def get_cpu_count() -> int:
             return 1
-        return self.get_cpu_count()
+        # Maxing out number of cores used to 16 if more not requested to not overload on large machines
+        return min(16, self.get_cpu_count())
```
And we shouldn't be doing this silently, there should be warning being logged here IMO.
